Ruby make command to create a new post

jonatack opened the pull request:
This PR adds a Ruby make (rake) command to run from root or `/_posts`, that creates a new post file:

```
$ rake posts:new -- --help
Usage: rake posts:new -- <options>
    -p, --pr NUMBER                  Pull request number (required)
    -h, --host USERNAME              Host's GitHub username (required)
    -d, --date YYYY-MM-DD            Meeting date in ISO8601 format (optional, defaults to next Wednesday)
    -H, --help                       Display this help

$ rake posts:new --pr 16402 --host jnewbery
New post file _posts/2019-10-09-#16402.md created successfully.
```
Args are order-independent.

`--pr` and `--host` are required and an error is signalled if they are omitted.

If no date is passed, it is automatically set to the upcoming Wednesday and a message is printed.

If the file already exists:
```
$ rake posts:new -- --pr 16402 --host satoshi
Filename _posts/2019-10-09-#16402.md already exists. Nothing done, exiting.
```
If the user supplies an invalid date:
```
$ rake posts:new -- --pr 16402 --host satoshi --date 20-11-12 
Error: Invalid date (20-11-12 received, YYYY-MM-DD needed). Nothing done, exiting.
```
If the given PR does not exist:
```
rake posts:new -- --date 2019-10-09 --host harding --pr 16909
Error: HTTP 404 Not Found. PR 16909 doesn't exist.
```

Additional metadata was added; to be adjusted as needed. Also basic removal of a few unwanted tags, see the `UNDESIRED_LABELS` array ('refactoring', 'needs conceptual review', 'needs rebase', 'waiting for author').

Sample markdown file generated in `_posts/2019-10-09-#16409.md`
```markdown
---
layout: pr
date: 2019-10-09
title: "Remove mempool expiry, treat txs as replaceable instead"
pr: 16409
authors: [MarcoFalke]
components: ["mempool", "tx fees and policy"]
host: jachiang
---

## Notes

## Questions
```

jonatack commented:
Proposed to-dos:

- Add a `--date=YYYY-MM-DD` argument, and date info in the help.

- Add a rake task to update missing metadata for all posts (as needed for #39). I'm thinking of a final user interface like:
```
rake posts:new --pr=id --host=name --date(optional)=YYYY-MM-DD
rake posts:update --pr=<id or [ids]> --all
rake posts:update_all
```

jonatack commented:
Pushed a few improvements pertaining to the meeting date handling. This PR should be ready now. Will make an eventual follow-up PR for the `rake posts:update` command to update missing metadata in existing posts if this PR is good.

jnewbery commented on Gemfile:
```diff
@@ -2,8 +2,17 @@
 
 source "https://rubygems.org"
 
+git_source(:github) { |repo_name| "https://github.com/#{repo_name}.git" }
```
What's this used for?

jonatack replied:
This ensures https is used when pulling in dependencies from a tag, a branch, or master rather than a gem release, which I was initially doing with curb, but not now. Removed.

jnewbery commented on Rakefile:
```diff
@@ -0,0 +1,128 @@
+    pr, host, date = get_cli_options
+
+    # Set default values for host and date if not supplied by user.
+    host ||= 'jnewbery'
```
I suggest leaving this as blank if not provided. The only person that this default could possibly benefit would be me, so it doesn't seem worthwhile adding it.

jonatack replied:
Done.

jnewbery commented on Rakefile:
```diff
@@ -0,0 +1,128 @@
+    line.puts "authors: [#{response.dig('user', 'login')}]"
+    line.puts "components: #{parse_components(response['labels'])}"
+    line.puts "created_at: \"#{response["created_at"]}\""
+    line.puts "state: #{response['state']}"
```
I'd prefer not to add this (and the `merged` field below), since those properties change. The review club page would become outdated if the state/merged field changed after this script was run the first time.

jonatack replied:
Done.

jnewbery commented on Rakefile:
```diff
@@ -0,0 +1,128 @@
+    line.puts "created_at: \"#{response["created_at"]}\""
+    line.puts "state: #{response['state']}"
+    line.puts "merged: #{response['merged']}"
+    line.puts "body: \"#{response['body'].split("\r").first}\""
```
Remove this too. There's no guarantee that the first line of the body is anything useful.

jonatack replied:
Done.

jnewbery commented on Rakefile:
```diff
@@ -0,0 +1,128 @@
+    line.puts "pr: #{response['number']}"
+    line.puts "authors: [#{response.dig('user', 'login')}]"
+    line.puts "components: #{parse_components(response['labels'])}"
+    line.puts "created_at: \"#{response["created_at"]}\""
```
Are you planning to use this data somewhere? If not, lets not add it yet.

jonatack replied:
Done.

jnewbery commented on README.md:
```diff
@@ -13,16 +13,20 @@ You'll need [Ruby & Jekyll](https://jekyllrb.com/docs/installation/) to run the
+initial Notes and Questions headers.
+
+You may need to run `gem install curb` the first time to install Ruby libcurl
+bindings.
```
It's unfortunate that this might also involve installing libcurl. For example, on my ubuntu machine I needed to run `sudo apt install libcurl4-openssl-dev` before running `gem install curb`.

jonatack replied:
For a long time, Curb had fewer dependencies and was more minimal and fast than other Ruby HTTP libraries. I haven't checked in the last 2-3 years though, so I'll have another look.

jonatack replied:
For a simple use case like this, no HTTP client is needed. I reimplemented the HTTP GET call without Curb/libcurl or any other external lib.

jonatack commented:
Thanks for the review, John! I added a type error bug in my last force push that caused the issues you found -- good eye. It was a missing `to_s` to convert the default meeting date to a string before splitting it in the validation. Thanks for your feedback on the other items you mentioned -- done except for libcurl, can have a look at that.

jnewbery commented on Rakefile:
```diff
@@ -0,0 +1,123 @@
+def get_cli_options(options = {})
+  OptionParser.new do |opts|
+    opts.banner = 'Usage: rake posts:new -- <options>'
+    opts.on('--pr NUMBER', 'Pull request number') { |pr| options[:pr] = pr }
```
Is it possible to make this argument required and exit with a helpful error and print out the help text.

Currently, the error is not very helpful:

```
→ rake posts:new
HTTP 404 Not Found
{"message":"Not Found","documentation_url":"https://developer.github.com/v3"}
```

jonatack replied:
Agreed, was about to push this:

"Error: Missing --pr argument (required). Run `rake posts:new -- --help` for help info."

jnewbery commented on Rakefile:
```diff
@@ -0,0 +1,123 @@
+    pr, host, date = get_cli_options
+
+    # Set default value for meeting date if not supplied by user.
+    date ||= next_wednesday.to_s
```
It'd be nice if this printed out a message to the user like `date set to next Wednesday: <date>`

jonatack replied:
Good idea. Done.

jnewbery commented on Rakefile:
```diff
@@ -0,0 +1,123 @@
+  OptionParser.new do |opts|
+    opts.banner = 'Usage: rake posts:new -- <options>'
+    opts.on('--pr NUMBER', 'Pull request number') { |pr| options[:pr] = pr }
+    opts.on('--host USERNAME', "Host's GitHub username") { |host| options[:host] = host }
```
This should print a warning (or error?) if no host is provided.

jonatack replied:
Yes. Adding:

    Error: Missing --host argument (required). Run `rake posts:new -- --help` for help info.

jnewbery commented on Rakefile:
```diff
@@ -0,0 +1,133 @@
+def get_cli_options(options = {})
+  OptionParser.new do |opts|
+    opts.banner = 'Usage: rake posts:new -- <options>'
+    opts.on('--pr NUMBER', 'Pull request number') { |pr| options[:pr] = pr }
```
nit: update help text to say that this is required

jonatack replied:
Good idea, done.

jnewbery commented on Rakefile:
```diff
@@ -0,0 +1,133 @@
+  OptionParser.new do |opts|
+    opts.banner = 'Usage: rake posts:new -- <options>'
+    opts.on('--pr NUMBER', 'Pull request number') { |pr| options[:pr] = pr }
+    opts.on('--host USERNAME', "Host's GitHub username") { |host| options[:host] = host }
```
nit: update help text to say that this is required

jonatack replied:
Done.

jonatack commented:
Good ideas. Squashed the review feedback changes to one commit and updated the PR description.

jonatack commented:
For info, we can also have short args as follows if you prefer:
```
$ rake posts:new -- --help
Usage: rake posts:new -- <options>
    -p, --pr NUMBER                  Pull request number (required)
    -h, --host USERNAME              Host's GitHub username (required)
    -d, --date YYYY-MM-DD            Meeting date in ISO8601 format (optional, defaults to next Wednesday)
    -H, --help                       Display this help
```

jnewbery commented:
> we can also have short args

yes please!

jonatack commented:
Added short args and updated the README.

jonatack commented:
Added short args, removed the Curb/libcurl dependencies, improved the case when a 404 is returned

```
$ rake posts:new -- -p 53319 -h jon
Date set to next Wednesday: 2019-10-09
Error: HTTP 404 Not Found. PR 53319 doesn't exist.
```
and the case when a PR contains non-numerical characters
```
$ rake posts:new -- -p 164E01P -h jon 
Error: Non-numerical PR 164E01P received. Nothing done, exiting.
```
and squashed the commits.

jnewbery commented:
Doesn't work for me :(

```
→ rake posts:new 15921
rake aborted!
NameError: uninitialized constant OpenSSL
Did you mean?  OpenStruct
/home/john/Code/docs/bitcoin-core-review-club/Rakefile:25:in `<top (required)>'
/home/john/.rvm/gems/ruby-2.5.1/gems/rake-13.0.0/exe/rake:27:in `<top (required)>'
/home/john/.rvm/gems/ruby-2.5.1/bin/ruby_executable_hooks:24:in `eval'
/home/john/.rvm/gems/ruby-2.5.1/bin/ruby_executable_hooks:24:in `<main>'
(See full trace by running task with --trace)
→ rake posts:new 15921 --trace
rake aborted!
NameError: uninitialized constant OpenSSL
Did you mean?  OpenStruct
/home/john/Code/docs/bitcoin-core-review-club/Rakefile:25:in `<top (required)>'
/home/john/.rvm/gems/ruby-2.5.1/gems/rake-13.0.0/lib/rake/rake_module.rb:29:in `load'
/home/john/.rvm/gems/ruby-2.5.1/gems/rake-13.0.0/lib/rake/rake_module.rb:29:in `load_rakefile'
/home/john/.rvm/gems/ruby-2.5.1/gems/rake-13.0.0/lib/rake/application.rb:703:in `raw_load_rakefile'
/home/john/.rvm/gems/ruby-2.5.1/gems/rake-13.0.0/lib/rake/application.rb:104:in `block in load_rakefile'
/home/john/.rvm/gems/ruby-2.5.1/gems/rake-13.0.0/lib/rake/application.rb:186:in `standard_exception_handling'
/home/john/.rvm/gems/ruby-2.5.1/gems/rake-13.0.0/lib/rake/application.rb:103:in `load_rakefile'
/home/john/.rvm/gems/ruby-2.5.1/gems/rake-13.0.0/lib/rake/application.rb:82:in `block in run'
/home/john/.rvm/gems/ruby-2.5.1/gems/rake-13.0.0/lib/rake/application.rb:186:in `standard_exception_handling'
/home/john/.rvm/gems/ruby-2.5.1/gems/rake-13.0.0/lib/rake/application.rb:80:in `run'
/home/john/.rvm/gems/ruby-2.5.1/gems/rake-13.0.0/exe/rake:27:in `<top (required)>'
/home/john/.rvm/gems/ruby-2.5.1/bin/rake:23:in `load'
/home/john/.rvm/gems/ruby-2.5.1/bin/rake:23:in `<main>'
/home/john/.rvm/gems/ruby-2.5.1/bin/ruby_executable_hooks:24:in `eval'
/home/john/.rvm/gems/ruby-2.5.1/bin/ruby_executable_hooks:24:in `<main>'
```

jonatack commented:
I'll install Ruby 2.5.1 and see. Perhaps a missing library. I'm running Ruby 2.6.5 and 2.7 locally.

jonatack commented:
Not reproduced on a fresh install of 2.5.1 followed by `gem install bundler` and `bundle update`.
```
~/projects/bitcoin/review-club (create-post-script)$ ruby -v
ruby 2.5.1p57 (2018-03-29 revision 63029) [x86_64-linux]

~/projects/bitcoin/review-club (create-post-script)$ rake posts:new 15921
Error: Missing required --pr argument. Run `rake posts:new -- --help` for info.
```

That said, I'll just remove that OpenSSL line, it's not a likely error.
```diff
   Net::ProtocolError,
-  Timeout::Error,
-  OpenSSL::SSL::SSLError
+  Timeout::Error
```

jonatack commented:
Removed. Would you mind testing it again?

jnewbery commented:
tACK 8f5db7514ba2f82b922ce19eb4ee2136fa643482.

Tested:
- no PR argument (fails)
- non-numeric PR argument (fails)
- numeric PR argument that doesn't match a PR (fails)
- no host argument (fails)
- invalid date argument (fails)
- duplicate PR (fails)
- valid PR and host arguments, no date argument (succeeds)
- valid PR host and date arguments (succeeds)

Great stuff. Thanks Jon!

jonatack commented:
> Tested:
> 
>     * no PR argument (fails)
> 
>     * non-numeric PR argument (fails)
> 
>     * numeric PR argument that doesn't match a PR (fails)
> 
>     * no host argument (fails)
> 
>     * invalid date argument (fails)
> 
>     * duplicate PR (fails)
> 
>     * valid PR and host arguments, no date argument (succeeds)
> 
>     * valid PR host and date arguments (succeeds)
> 
> 
> Great stuff. Thanks Jon!

Excellent list of missing tests! Thanks :smiley:
